### PR TITLE
Call an optional cleanup routine before a node is destroyed

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
I've realized what was causing the segfaults in our Agora test-suite. It's a race condition between these calls:

```
// e.g. calls SQL shutdown methods which cannot be called in the dtor
node.shutdown();  

// race condition *here*

// destroys Node, shuts down event loop
node.ctrl.shutdown();
```

The logging confirmed it.

I've implemented a simple fix in this PR. A nice plus is that `shutdown` no longer has to be an `API` endpoint, and we can remove it from our interfaces in Agora.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
tl;dr: We will replace `node.shutdown && node.ctrl.shutdown` with just `node.ctr.shutdown` in Agora, and also remove `shutdown` from its API.

**Comment by AndrejMitrovic** [comment]
Fixed up the implementation according to tips by @Geod24.

Note that I couldn't use the full type from the `spawn` call because this type is only defined in the instantiation of `spawned!(Implementation)` and it's not available in the outer `RemoteAPI!(API)`. So client code will have to downcast if they wish to call methods not in the API.

File: source/geod24/LocalRest.d [hunk 1]
```diff
@@ -393,6 +398,13 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
         }
     }
 
+    // Ask the node to shut down. Kept here as we need access to `API` symbol.
+    private static struct ShutdownCommand
+    {
+        /// Any callback to call before the Node's destructor is called
+        void function (API) callback;
```
Comment (AndrejMitrovic) [inline source/geod24/LocalRest.d]: I deliberately omitted `nothrow` because exceptions are already caught and logged in LocalRest. So it's nicer UX.

File: source/geod24/LocalRest.d [hunk 2]
```diff
@@ -40,6 +40,16 @@
     methods, for some defined time or until unblocked.
     See `sleep`, `filter`, and `clearFilter` for more details.
 
+    Shutdown:
+    The control interface has a shutdown method. It may be called when one
+    wants to shut down the node and call its destructor. Additionally some
+    classes may wish to do resource deallocation separately from their
+    destructors, in particular when those methods themselves allocate
+    (D destructors may not allocate GC memory). Therefore, you may provide
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: That is actually not a concern for localrest. The node is allocated using `scope`, which means it's stack allocated, which means the destructor runs deterministically and can allocate.

File: source/geod24/LocalRest.d [hunk 3]
```diff
@@ -40,6 +40,16 @@
...
+    a shutdown routine in the call to `shutdown`, it will be called before
+    the Node is destroyed. This `shutdown` method should not schedule any
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Why The Capitalization Of Node ?

File: source/geod24/LocalRest.d [hunk 4]
```diff
@@ -115,11 +125,6 @@ private struct TimeCommand
     bool drop = false;
 }
 
-/// Ask the node to shut down
-private struct ShutdownCommand
```
Comment (Geod24) [inline source/geod24/LocalRest.d:119]: Why did you move it ? Just template it.

File: source/geod24/LocalRest.d [hunk 5]
```diff
@@ -823,6 +847,46 @@ unittest
...
+
+    static interface API
+    {
+        @safe:
+        public @property ulong pubkey ();
+    }
+
+    static class MockAPI : API
+    {
+        @safe public override @property ulong pubkey ()
```
Comment (Geod24) [inline source/geod24/LocalRest.d]:
```suggestion
        public override @property ulong pubkey () @safe
```

File: source/geod24/LocalRest.d [hunk 6]
```diff
@@ -823,6 +847,46 @@ unittest
...
+        auto mock = cast(MockAPI)node;
+        assert(mock !is null);
+        mock.shutdown();
+        onDestroy_called = true;
+    }
+
+    scope test = RemoteAPI!API.spawn!MockAPI();
+    assert(test.pubkey() == 42);
+    test.ctrl.shutdown(&onDestroy);
+    core.thread.Thread.sleep(1.seconds);  // ctr.shutdown call is asynchronous
```
Comment (Geod24) [inline source/geod24/LocalRest.d]:
```suggestion
    while (!core.atomic.atomicLoad(dtor_called)) core.atomic.pause();
```

The problem is that this test doesn't fail, but at least it won't spuriously fail.

File: source/geod24/LocalRest.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d:119]: I thought if I templatized it you would complain about there being too many templates. :D

Comment (Geod24) [reply source/geod24/LocalRest.d:119]: Well it's a template anyway 🤷

File: source/geod24/LocalRest.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Ah well it was just an example, I could remove it. Anyway how is a destructor supposed to know if the object it's cleaning is allocated on the heap vs the stack?

https://github.com/biozic/d2sqlite3/blob/0280ef1769f0e0907980a241b9ac68e444a1fbd5/source/d2sqlite3/database.d#L89

File: source/geod24/LocalRest.d (hunk 1 above)
Comment (Geod24) [reply source/geod24/LocalRest.d]: Good point

Comment (Geod24) [reply source/geod24/LocalRest.d]: Maybe document what happens if one `throw` then ?

File: source/geod24/LocalRest.d (hunk 2 above)
Comment (Geod24) [reply source/geod24/LocalRest.d]: Looking at this method, I thought: "Please don't implement this by catching `InvalidMemoryOperationError`... PLEASE".
[Welp](https://github.com/biozic/d2sqlite3/blob/08e77f5f2da3ead57d62c80226bd85e41c27fbd7/source/d2sqlite3/internal/memory.d#L70)

File: source/geod24/LocalRest.d (hunk 6 above)
Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: > core.atomic.pause

hello there

Cool. Didn't know about this.

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: We should fix other places where we have sleep calls like this.

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Is `core.atomic.pause` a new feature? I'm getting a build failure on ldc 14 and DMD 2.085.1

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Oh those are quite old compilers too. But I assume you still want to support them.

File: source/geod24/LocalRest.d [hunk 7]
```diff
@@ -40,6 +40,14 @@
     methods, for some defined time or until unblocked.
     See `sleep`, `filter`, and `clearFilter` for more details.
 
+    Shutdown:
+    The control interface has a shutdown method. You may provide a
+    shutdown routine in the call to `shutdown`, it will be called before
+    the node is destroyed. This `shutdown` method should not schedule any
+    tasks / fibers within the node, otherwise it may introduce race conditions.
+    If this routine throws, LocalRest will log the error in the console and
+    proceed with destroying the stack-allocated node.
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: I gave a shot at rewriting it:
```suggestion
    Shutdown:
    The control interface has a shutdown method that can be used to terminate
    a node gracefully. When the shutdown request is handled by the node,
    the event loop will exit and the thread will terminate. While the destructor
    of the node will be called, it might not usable for some actions, for example
    because D destructors may not allocate GC memory, or one may want
    to perform some test-specific operations, such a logging some data in case of failure.
    Therefore, you may provide a shutdown routine in the call to `shutdown`.
    It must accept a single argument of the interface type, and will be called
    with the implementation object just before the node is destroyed.
    If this routine throws, LocalRest will log the error in the console and
    proceed with destroying the stack-allocated node.
    Note that control requests are asynchronous, hence requests from the node
    might be processed / send by the node until the request is actually processed.
```

I also think we should add a mention that all control interface methods are asynchronous (in the previous paragraph). I also wanted to focus more on the shutdown method itself than the optional argument, while keeping the limitations that brought us here and you previously documented.

**Comment by Geod24** [comment]
You might need to update the line breaks tho :P

File: source/geod24/LocalRest.d (hunk 6 above)
Comment (Geod24) [reply source/geod24/LocalRest.d]: Yeah a `sleep` in a loop might just be simpler

Comment (Geod24) [reply source/geod24/LocalRest.d]: My main concern was the lack of a retry mechanism

**Comment by AndrejMitrovic** [comment]
> You might need to update the line breaks tho :P

I think it looks okay.

**Review by Geod24 — APPROVED** [review]
